autoprotocol/informatics.py

```diff
@@ -0,0 +1,119 @@
+"""
+Contains all the Autoprotocol Informatics objects
+
+    :copyright: 2020 by The Autoprotocol Development Team, see AUTHORS
+        for more details.
+    :license: BSD, see LICENSE for more details
+
+"""
+
+from abc import abstractmethod
+
+from .compound import Compound
+from .container import Container, WellGroup
+from .util import is_valid_well
+
+
+class Informatics:
+    """
+    Base class for informatics attribute in an instruction that is to later be
+    encoded as JSON.
+    """
+
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def as_dict(self):
+        pass
+
+
+class AttachCompounds(Informatics):
+    """
+    informatics type attach_compounds is constructed as a list of dict detailing
+    new compounds
+
+    Parameters
+    ----------
+    data:
+        dict of informatics data
+    all_wells: WellGroup
+        All wells used in the instruction that are applicable for attach_compounds
+        to take effect on.
+
+    Returns
+    -------
+    AttachCompounds
+
+    Raises
+    ------
+    TypeError
+        data must be a dict
+    TypeError
+        wells must be Well, list of Well, or WellGroup
+    ValueError
+        wells must be part of wells or containers instruction operates on
+    TypeError
+        compounds must be a list
+    TypeError
+        each element in compounds must be Compound type
+    """
+
+    def __init__(self, data: dict, all_wells):
```

Comment:
:+1: Thanks for adding types. We should look to eventually type annotate everything and use a static-type checker like mypy.

Reply:
On more thought... it feels like we could simplify this signature? I.e. Specify `wells` and `compounds` directly, rather than using this `data` concept.

Reply:
Hmm yea.. this is strange. Ideally, we should simplify this such that we don't have to pass `all_wells` directly and only validate this when we're serializing the Instruction.

Reply:
`data` - I think we still want to have 'data' to be the common parameter across all types so they are modular, and can be deserialized automatically. I think we still want `data` to be the common parameter to do so? (unless we specify in Protocol for this type, these are arguments, but Protocol shouldn't have to know that?) The parameters within 'data' and validation could be different, and those can be defined in each class.
`all_well` - ya, thought it was weird, too... :p moved this check to outside.